### PR TITLE
fix null byte handling in JSON paths

### LINKED ISSUE #2340: Panic on document of death
Experienced on airmail:

Apr 2 10:40:24 quickwit-indexer-14 quickwit-indexer thread 'blocking-5' panicked at /usr/local/cargo/git/checkouts/tantivy-f70b7ea03dadae9a/4e79e11/src/schema/term.rs:318:42:

### PR DESCRIPTION
**Pull request by PSeitz**
closes https://github.com/quickwit-oss/tantivy/issues/2340

### CODE REVIEW

File: columnar/src/columnar/writer/serializer.rs [hunk 1]
```diff
@@ -18,7 +18,12 @@ pub struct ColumnarSerializer<W: io::Write> {
 /// code.
 fn prepare_key(key: &[u8], column_type: ColumnType, buffer: &mut Vec<u8>) {
     buffer.clear();
-    buffer.extend_from_slice(key);
+    // Convert 0 bytes to '0' string, as 0 bytes are reserved for the end of the path.
+    if key.contains(&0) {
```
Comment (fulmicoton) [inline columnar/src/columnar/writer/serializer.rs]:
```suggestion
    if key.contains(&0u8) {
```

File: src/schema/term.rs [hunk 2]
```diff
@@ -218,6 +218,21 @@ impl Term {
         &mut self.0[len_before..]
     }
 
+    /// Appends json path bytes to the Term.
+    ///
+    /// This function returns the segment that has just been added.
+    #[inline]
+    pub fn append_path(&mut self, bytes: &[u8]) -> &mut [u8] {
+        let len_before = self.0.len();
+        if bytes.contains(&0) {
```
Comment (fulmicoton) [inline src/schema/term.rs]:
```suggestion
        if bytes.contains(&0u8) {
```

File: src/schema/term.rs [hunk 3]
```diff
@@ -218,6 +218,21 @@ impl Term {
         &mut self.0[len_before..]
     }
 
+    /// Appends json path bytes to the Term.
```
Comment (fulmicoton) [inline src/schema/term.rs:221]: can you comment on the special behavior if bytes contains 0u8

File: columnar/src/columnar/writer/serializer.rs [hunk 4]
```diff
@@ -102,7 +107,7 @@ mod tests {
         let mut buffer: Vec<u8> = b"somegarbage".to_vec();
         prepare_key(b"root\0child", ColumnType::Str, &mut buffer);
         assert_eq!(buffer.len(), 12);
-        assert_eq!(&buffer[..10], b"root\0child");
```
Comment (fulmicoton) [inline columnar/src/columnar/writer/serializer.rs:105]: which sep do we use for fields in columnar? is it \1?

**Review by fulmicoton — APPROVED** [review]

File: columnar/src/columnar/writer/serializer.rs (hunk 4 above)
Comment (PSeitz) [reply columnar/src/columnar/writer/serializer.rs:105]: yes `\u{1}`

**Pull request merged**